### PR TITLE
UDS over DoIP

### PR DESCRIPTION
**Pull request by Cr0wTom**
New module targeting UDSonIP (ISO 14229-5). 

The module includes all the functionality from the UDS module and the seed_randomness_fuzzer from the uds_fuzz module. For this instance, a significant amount of code was borrowed from the UDS module. 

Additionally, udsoncan (https://github.com/pylessard/python-udsoncan) and doipclient (https://pypi.org/project/doipclient/) libraries were used.

### CODE REVIEW

**Review by kasperkarlsson — CHANGES_REQUESTED** [review]
Looking forward to merge this! I just noted some small issues during review, which would be nice to get fixed before merging :+1: Just let me know if any of the comments requires a clarification.

File: tool/modules/doip.py [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+        
+    except OSError:
+        print("Please check the connection and try again.\n")
+
+
+def seed_randomness_fuzzer(args):
+    """Wrapper used to initiate security randomness fuzzer"""
+    arb_id_request = args.src
+    arb_id_response = args.dst
+    reset_type = args.reset
```
[the remark this answers is not in the document]
Comment (Cr0wTom) [reply tool/modules/doip.py:666]: Done, just a typo :)

File: tool/modules/doip.py [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+    elif reset_type == 2:
+        client.ecu_reset(ECUReset.ResetType.keyOffOnReset)
+    elif reset_type == 3:
+        client.ecu_reset(ECUReset.ResetType.softReset)
+    elif reset_type == 4:
+        client.ecu_reset(ECUReset.ResetType.enableRapidPowerShutDown)
+    elif reset_type == 5:
+        client.ecu_reset(ECUReset.ResetType.disableRapidPowerShutDown)
+
+def extended_session(client, session_type):
```
[the remark this answers is not in the document]
Comment (Cr0wTom) [reply tool/modules/doip.py:98]: Basically I implemented it because there are a couple of places where the user can specify an option for the diagnostic session he/she needs (e.g. 319, 514, 632, etc.). This means that this exact code of the extended_session() should be copied more than one time inside the code which from my perspective makes the code less clean. Let me know your preference and I will resolve it :)

File: tool/modules/doip.py [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+        pass
+
+# Duplicate testing from https://www.iditect.com/guide/python/python_howto_find_the_duplicates_in_a_list.html
+def find_duplicates(sequence):
+    first_seen = set()
+    first_seen_add = first_seen.add
+    duplicates = set(i for i in sequence if i in first_seen or first_seen_add(i))
+    return duplicates
+
+def ecu_reset(client, reset_type):
```
[the remark this answers is not in the document]
Comment (Cr0wTom) [reply tool/modules/doip.py:86]: Same as my answer for extended_session

File: tool/modules/doip.py [hunk 4]
```diff
@@ -0,0 +1,973 @@
...
+            
+            doip_client = DoIPClient(ip, logical_address, client_logical_address  = arb_id_response )
+        
+            conn = DoIPClientUDSConnector(doip_client)
+            
+            s = struct.pack('<h', service_id) 
+            
+            doip_message = struct.pack("!HH", arb_id_response, arb_id_request) + s[:1] + b'\x00'
+            
+            payload_type = 0x8001
```
[the remark this answers is not in the document]
Comment (Cr0wTom) [reply tool/modules/doip.py:315]: solved in line 74. added reference :)

File: tool/modules/doip.py [hunk 5]
```diff
@@ -0,0 +1,973 @@
...
+                        response_service_id = response[2]
+                        status = response[2]
+                        if response_id != Constants.NR_SI:
+                            request_id = service_id
+                            found_services.append(request_id)
+                        elif status != NegativeResponseCodes.SERVICE_NOT_SUPPORTED:
+                            # Any other response than "service not supported" counts
+                            found_services.append(response_service_id)
+                    
+            except ConfigError:
```
[the remark this answers is not in the document]
Comment (Cr0wTom) [reply tool/modules/doip.py:341]: Basically this has to do with 2 things:

- I could not disable udsoncan logger, so I was getting big error messages that wasn't helping somehow
- A lot of exceptions are just handling of behavior that was needed to enumerate the unit. for example I was getting the configerror for the first 10-20 services of the 0x00-0xff range, and then it was working properly. I could only fix it and burry the error messages (false positives probably) with the configerror exception.

Basically, with the code you see I managed to get uncluttered results, with meaningful errors, and avoiding the meaningless errors. As I said in my previous module, I am not a developer, so it would be nice if you can propose a more elegant solution :)

File: tool/modules/doip.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kasperkarlsson) [reply tool/modules/doip.py:102]: I see - let's keep it like this, then! Thanks for explaining 👍

File: tool/modules/doip.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kasperkarlsson) [reply tool/modules/doip.py:90]: Same answer here as well - let's keep it like this 👍

File: tool/modules/doip.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kasperkarlsson) [reply tool/modules/doip.py:341]: Seeing as you have a motivation for this pattern and I cannot setup a suitable test environment, I don't really dare to suggest any exact changes in the code for exceptions which I cannot trigger. Let's keep it like this.
